### PR TITLE
transactions: Amend pending transactions

### PR DESCRIPTION
**Pull request by kennylevinsen**
The transaction system contains a necessary optimization where a popped
transaction is combined with later, similar transactions. This breaks
the chronological order of states, and can lead to desynchronized
geometries.

To fix this, we replace the queue with only 2 transactions: current and
pending. If a pending transaction exists, it is updated with new state
instead of creating additional transactions.

As we never have more than a single waiting transaction, we no longer
need the queue optimization that is causing problems.

Closes: https://github.com/swaywm/sway/issues/6012

### CODE REVIEW

File: sway/desktop/transaction.c [hunk 1]
```diff
@@ -162,14 +178,33 @@ static void copy_container_state(struct sway_container *container,
...
+
+	// Check if we have an instruction for this node already, in which case we
+	// update that.
+	if (node->ntxnrefs > 0) {
+		for (int idx = 0; idx < transaction->instructions->length; idx++) {
+			struct sway_transaction_instruction *other =
+				transaction->instructions->items[idx];
+			if (other->node == node) {
+				instruction = other;
+				goto update;
```
Comment (Xyene) [inline sway/desktop/transaction.c:191]: This can be a `break`, and the code below can be within an `if (!instruction)` block, avoiding a `goto`.

File: sway/desktop/transaction.c [hunk 2]
```diff
@@ -533,20 +527,31 @@ void transaction_commit_dirty(void) {
 	if (!server.dirty_nodes->length) {
 		return;
 	}
-	struct sway_transaction *transaction = transaction_create();
-	if (!transaction) {
-		return;
+
+	struct sway_transaction *transaction = NULL;
+	if (server.transactions->length > 1) {
```
Comment (Xyene) [inline sway/desktop/transaction.c:532]: Hmm, I get that doing things like this gets to reuse the dequeuing code, but rather than an array of length 2, I think this would be more descriptive as two separate fields: `server.current_transaction` and `server.pending_transaction` (or whatever naming matches what we do for double-buffering state elsewhere).

**Review by Xyene — COMMENTED** [review]
Thanks for working on this! The overall approach seems reasonable to me, and I've been running this without being able to induce issues.

File: sway/desktop/transaction.c (hunk 1 above)
Comment (kennylevinsen) [reply sway/desktop/transaction.c]: I like the goto better, but oh well. Changed.

File: sway/desktop/transaction.c (hunk 2 above)
Comment (kennylevinsen) [reply sway/desktop/transaction.c]: Agreed, I initially kept the the same just for change simplicity. Updated.

**Review by Xyene — APPROVED** [review]
This looks good to me and appears to work in all cases I've tried, thanks.

@emersion, @RPigott -- thoughts?

**Comment by emersion** [comment]
Obligatory bikeshed: `current_transaction` isn't really "current", right? Throughout wlroots, "current" means currently displayed on screen, aka. state used for rendering. But as far as I understand there are two transactions:

- A "pending" transaction, currently being constructed.
- A transaction which has sent `configure` events to clients, but which is still waiting for client ACKs.

Maybe we can rename the latter to "queued"? That's the wording used in the DRM backend when the state has been submitted but not yet applied.

**Comment by kennylevinsen** [comment]
> Obligatory bikeshed:

:)

> Maybe we can rename the latter to "queued"? That's the wording used in the DRM backend when the state has been submitted but not yet applied.

Sure. current/pending did give a bit of a double-buffering vibe, which indeed isn't what is going on here.

File: sway/desktop/transaction.c [hunk 3]
```diff
@@ -472,6 +459,17 @@ static void transaction_commit(struct sway_transaction *transaction) {
 	}
 }
 
+static void transaction_commit_pending() {
```
Comment (emersion) [inline sway/desktop/transaction.c:462]: Nit: `(void)`

File: sway/desktop/transaction.c [hunk 4]
```diff
@@ -104,8 +108,16 @@ static void copy_workspace_state(struct sway_workspace *ws,
 	state->layout = ws->layout;
 
 	state->output = ws->output;
-	state->floating = create_list();
-	state->tiling = create_list();
+	if (state->floating) {
+		state->floating->length = 0;
```
Comment (emersion) [inline sway/desktop/transaction.c:112]: Why are these changes now necessary? Is it just a bugfix for leaked memory?

If so, please split these into a separate commit.

Comment (emersion) [reply sway/desktop/transaction.c:112]: Ah, my bad, this is required for the new `transaction_add_node` logic.

File: include/sway/server.h [hunk 5]
```diff
@@ -86,7 +88,8 @@ struct sway_server {
 	struct wlr_foreign_toplevel_manager_v1 *foreign_toplevel_manager;
 
 	size_t txn_timeout_ms;
-	list_t *transactions;
+	struct sway_transaction *current_transaction;
+	struct sway_transaction *pending_transaction;
```
Comment (emersion) [inline include/sway/server.h:92]: Maybe add comments to explain what these are

**Comment by emersion** [comment]
Overall looks good!

File: include/sway/server.h (hunk 5 above)
Comment (kennylevinsen) [reply include/sway/server.h:102]: Added comments for the 4 transaction-related fields.

**Review by emersion — APPROVED** [review]
LGTM, thanks a lot!
